src/java.base/share/classes/sun/security/ssl/SSLBasicKeyDerivation.java

```diff
@@ -28,29 +28,33 @@
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.security.GeneralSecurityException;
-import java.security.spec.AlgorithmParameterSpec;
+import javax.crypto.KDF;
 import javax.crypto.SecretKey;
+import javax.crypto.spec.HKDFParameterSpec;
 import javax.net.ssl.SSLHandshakeException;
 
+import sun.security.ssl.CipherSuite.HashAlg;
+
 final class SSLBasicKeyDerivation implements SSLKeyDerivation {
-    private final String hashAlg;
+    private final String hkdfAlg;
     private final SecretKey secret;
     private final byte[] hkdfInfo;
+    private final int keyLen;
 
-    SSLBasicKeyDerivation(SecretKey secret, String hashAlg,
-            byte[] label, byte[] context, int length) {
-        this.hashAlg = hashAlg.replace("-", "");
+    SSLBasicKeyDerivation(SecretKey secret, HashAlg hashAlg, byte[] label,
+            byte[] context) {
+        this.hkdfAlg = hashAlg.hkdfAlgorithm;
         this.secret = secret;
-        this.hkdfInfo = createHkdfInfo(label, context, length);
+        this.hkdfInfo = createHkdfInfo(label, context, hashAlg.hashLength);
+        this.keyLen = hashAlg.hashLength;
     }
 
     @Override
-    public SecretKey deriveKey(String algorithm,
-            AlgorithmParameterSpec keySpec) throws IOException {
+    public SecretKey deriveKey(String type) throws IOException {
         try {
-            HKDF hkdf = new HKDF(hashAlg);
-            return hkdf.expand(secret, hkdfInfo,
-                    ((SecretSizeSpec)keySpec).length, algorithm);
+            KDF hkdf = KDF.getInstance(hkdfAlg);
+            return hkdf.deriveKey(type,
+                    HKDFParameterSpec.expandOnly(secret, hkdfInfo, keyLen));
```

Comment:
I noticed we have done away with the `AlgorithmParameterSpec` parameter to this method. While the new implementation makes sense in light of the new parameter set, I wonder whether there was a deliberate use-case lost here... 

In the previous implementation, presumably, if the `keySpec`'s length was shorter than the expanded value, only a portion of the value would be used. With the new implementation, the hash algorithm's length of bytes is always used. 

I'm not sure how relevant this is, but it could be worth noting. If we had kept the old parameter, I suppose we could have used `deriveData` and truncated the result to `((SecretSizeSpec)keySpec).length` before returning a `SecretKey`.

Reply:
Searching among the JSSE code base, it looks that `SSLBasicKeyDerivation` class is only used in `Finished` class and the particular `SecretSizeSpec` value is always same as the current `keyLen` field. There is no other usage as far as I can tell. If we ended up using `SSLBasicKeyDerivation` class with a shorter key length, we can add an explicit `keyLen` argument to the `SSLBasicKeyDerivation` constructor. I see no need for the SecretSizeSpec class given the code flow in `Finished` class.